numpy/core/src/umath/reduction.c

```diff
@@ -54,7 +54,9 @@ allocate_reduce_result(PyArrayObject *arr, npy_bool *axis_flags,
 
     /* Build the new strides and shape */
     stride = dtype->elsize;
-    memcpy(shape, arr_shape, ndim * sizeof(shape[0]));
+    if (arr_shape != NULL && ndim > 0) {
```

Comment:
Why the check for ndim here?

Reply:
It seems to me it makes sense to copy only if both conditions are met. Perhaps more properly use `||`

Reply:
This seems to be normal arrays, so shape can only be NULL exactly when `ndim == 0`. So either check is definitely enough. If the NULL is what causes the tripping here, I guess we should go with that check, otherwise with the ndim one?

Reply:
I'd be inclined to create an npy_memcpy function, with semantics "if n is 0 src and/or dest may be null". That way, it's equivalent to the for loop you would write in place of a memcpy.

So for this line without such a helper, that would be `if (ndim)`. If ndim is non-zero but the pointer is still null, that's an internal error somewhere else, and something ubsan _should_ be flagging, not something we should be ignoring

Reply:
I agree with that.